tools/protodoc/protodoc.py

```diff
@@ -301,8 +311,15 @@ def FormatFieldAsDefinitionListItem(type_context, field):
   else:
     oneof_comment = ''
   anchor = FormatAnchor(FieldCrossRefLabel(type_context.name))
-  comment = '(%s) ' % FormatFieldType(
-      type_context, field) + type_context.LeadingCommentPathLookup()
+  annotations = []
+  if field.options.HasExtension(validate_pb2.rules):
+    rule = field.options.Extensions[validate_pb2.rules]
+    if rule.HasField('message'):
+      if rule.message.required:
```

Comment:
Does the annotation thing have a way of specifying the sometimes required stuff ? Or is it purely binary like optional/required?

Reply:
What does sometimes required mean? :)

Reply:
It means things that are required if some other option is set. I don't think we can reasonably capture this in the validator code in v1. All of that logic already exists in Envoy itself anyway as JSON schema doesn't check it either.